### ISSUE TITLE
Slower than single GPU

### ISSUE THREAD

**Issue by hankcs** [guotong1988/BERT-multi-gpu#3]
Hi, thanks for your work. I'd say that there is an error when measuring time. The unit of `time.time()` is second, not ms.

https://github.com/guotong1988/BERT-multi-gpu/blob/14c3ca9aaafe71e1bc816677ca36a12c81530b1d/run_pretraining_gpu.py#L726

Then, I would say that this naive `MirroredStrategy` is slower than single GPU. According to my experiments on `4`  Tesla V100 GPUs, this implementation takes around `88` seconds per 100 steps.

```
88.0555248260498  s
loss  22828.4058984375
------------
88.03012156486511  s
loss  17133.7074609375
------------
88.0766191482544  s
loss  13693.772515625
------------
88.04167246818542  s
loss  11408.907552083334
```

While the official BERT single-gpu version reports `2.9` steps per second, which means it takes around `34` seconds for `100` steps. Part of the log is as follows.

```
INFO:tensorflow:global_step/sec: 2.90046
INFO:tensorflow:examples/sec: 46.4073
```

You can insert a hook to record performance:

```
hook = tf.train.ProfilerHook(save_steps=100,
                                     output_dir=FLAGS.output_dir,
                                     show_memory=True)
estimator.train(input_fn=train_input_fn, max_steps=FLAGS.num_train_steps, hooks=[hook])
```

My experiment configuration is:

```
python3 run_pretraining_gpu.py \
  --input_file=data/tfrecord/sample.tfrecord \
  --output_dir=data/wiki_uncased_L-12_H-768_A-12 \
  --do_train=True \
  --do_eval=True \
  --bert_config_file=${BERT_BASE_DIR}/bert_config.json \
  --init_checkpoint=${BERT_BASE_DIR}/bert_model.ckpt \
  --train_batch_size=16 \
  --max_seq_length=256 \
  --max_predictions_per_seq=38 \
  --num_train_steps=1000 \
  --num_warmup_steps=10 \
  --learning_rate=2e-5
```

**Comment by guotong1988** [comment]
The official BERT also use Tesla V100 ? Thank you.

**Comment by guotong1988** [comment]
Have you eval the single GPU speed of this project ? Thank you.

**Comment by hankcs** [comment]
Yes, I ran both official BERT and your project on the same AWS instance, which has 4 Tesla V100. On single GPU, the project achieves `71` seconds.

```
71.15882539749146  s
loss  33404.156328125
------------
71.40574765205383  s
loss  22303.279544270834
```

**Comment by guotong1988** [comment]
Hi, I have changed some code and observe speedup 40% on my GPU.
Could you please eval on your machine? Thank you!

**Comment by eduOS** [comment]
I also found that multiple GPUs would reduce speed. A single GPU leads to faster speed but multiple GPUs can get bigger loss reduction.

**Comment by guotong1988** [comment]
Try https://github.com/guotong1988/BERT-multi-gpu/blob/master/run_pretraining_gpu_v2.py
?

**Comment by leonsim** [comment]
Hi, any update on this? Is the v2 script much more efficient than a single GPU?

**Comment by guotong1988** [comment]
50000step-4GPU-set8batchsize-3.75loss
50000step-1GPU-set8batchsize-6.00loss

**Comment by guotong1988** [comment]
global_step/sec: 2.05
300000 step
4 GPU
set 8 batchsize
loss ≈ 2

global_step/sec: 1.4
300000 step
1 GPU
set 16 batchsize
loss ≈ 2.2

global_step/sec: 2.6
300000 step
1 GPU
set 8 batchsize
loss ≈ 2.8

**Comment by guotong1988** [comment]
https://datascience.stackexchange.com/questions/43119/whats-the-advantage-of-multi-gpu-training-in-real

**Comment by guotong1988** [comment]
For `HierarchicalCopyAllReduce` `MirroredStrategy`, `global_step/sec` shows the sum of multi gpus' steps.
